flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServer.java

```diff
@@ -0,0 +1,411 @@
+					continue;
+				}
+				boolean updateOverview = false;
+				for (FileStatus jobArchive : jobArchives) {
+					Path jobArchivePath = jobArchive.getPath();
+					String jobID = jobArchivePath.getName();
+					if (cachedArchives.put(jobID, refreshDir) == null) {
+						try {
+							// contents of /:refreshDir/:jobid
+							FileStatus[] archiveFiles = fs.listStatus(jobArchivePath);
+							for (FileStatus archiveFile : archiveFiles) { 
+								if (archiveFile.isDir()) {
+									// contents of /:refreshDir/:jobid/jobs
+									for (FileStatus jobFile : fs.listStatus(archiveFile.getPath())) {
+										/**
+										 * src: /:refreshDir/:jobid/jobs/X
+										 * dst: /:webDir/jobs/X
+										 * 
+										 * X is either :jobid.json or the :jobid directory.
+										 */
+										FileCache.copy(jobFile.getPath(), new Path(webJobDir + "/" + jobFile.getPath().getName()), false);
+									}
+								} else {
+									/**
+									 * src: /:refreshDir/:jobid/joboverview.json
+									 * dst: /:webDir/joboverviews/:jobid.json
+									 */
+									FileCache.copy(archiveFile.getPath(), new Path(webOverviewDir, jobID + ".json"), false);
+									updateOverview = true;
+								}
+							}
+						} catch (IOException e) {
+							LOG.error("Failed to fetch archive file.", e);
+							cachedArchives.remove(jobID);
+							continue;
+						}
+					}
+				}
+				if (updateOverview) {
+					updateJobOverview(webDir);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Backup file fetcher that is used by the {@link HistoryServerStaticFileServerHandler} if it cannot find a file
+	 * within a job archive. This class does not share code with the {@link JobArchiveFetcherTask} since the former
+	 * fetches entire archives at a given location based on a path whereas this class only fetches one specific file
+	 * based on a REST URL.
+	 */
+	public static class JobFileFetcher {
+		private final Path webDir;
+		private final Map<String, Path> cachedArchives;
+
+		public JobFileFetcher(Path webDir, Map<String, Path> cachedArchives) {
+			this.webDir = webDir;
+			this.cachedArchives = cachedArchives;
+		}
+
+		public void fetchFileForRestRequest(String requestPath) throws IOException {
+			/**
+			 * Here we extract the job ID from the request path. We are either missing a file that resides
+			 * under /jobs/:jobid/* or /jobs/:jobid.json. In the latter case indexOf will
+			 * return -1 so we manually set 'to' to the index before '.json'.
+			 */
+			int from = 6;
+			int to = requestPath.indexOf('/', 7);
+			if (to == -1) {
+				to = requestPath.length() - 5;
+			}
+			String jobID = requestPath.substring(from, to);
+			Path sourcePath = new Path(new Path(cachedArchives.get(jobID), jobID), requestPath);
+			try {
+				FileCache.copy(sourcePath, new Path(webDir, requestPath), false);
+			} catch (Exception e) {
+				LOG.debug("Failed to retrieve file {} for job {}. This may indicate an incomplete or corrupt job archive.", sourcePath, e);
```

Comment:
I think the better pattern is 
```java
if (isDebug()) {
  LOG.error(msg, path, jobId, e);
} else {
  LOG.error(msg, path, jobId);
}
```
Furthermore, this line is missing the jobId replacement.